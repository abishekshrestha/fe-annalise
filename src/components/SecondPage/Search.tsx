import React, { SyntheticEvent } from "react";

interface IProps {
  name: string;
  search: string;
  imageList: any;
  loading: boolean;
  isValid: boolean;
  totalPages: number;
  page: number;
  handleSubmit?: (e: SyntheticEvent) => void;
  onPaginationClick?: () => void;
  onInputChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const Search = ({
  name = "",
  search = "",
  imageList = [],
  loading = false,
  isValid = true,
  totalPages = 0,
  page,
  handleSubmit,
  onPaginationClick,
  onInputChange,
}: IProps) => (
  <React.Fragment>
    <header className="header">
      <div className="container">
        <div className="row">
          <h1 className="col-12" id="name">
            Hello, {name}
          </h1>
        </div>
      </div>
    </header>
    <div className="container image-list">
      <div className="row">
        <div className="col-12">
          <form
            className="search-group"
            data-testid="search-form"
            onSubmit={handleSubmit}
          >
            <label htmlFor="search" hidden>
              Search
            </label>
            <input
              type="text"
              name="search"
              id="search"
              placeholder="Search"
              value={search}
              onChange={onInputChange}
              data-testid="inputbox"
            />
            <button type="submit">Search</button>
          </form>
          {!isValid && <div className="error">Please enter valid keyword</div>}
        </div>
      </div>

      {imageList.length > 0 && (
        <React.Fragment>
          <div className="row" data-testid="image-list">
            {imageList?.map((item: {}, index: number) => (
              // TODO: imageList and item type needs to be defined clearly
              //@ts-ignore
              <div className="col-4 col-12-sm" key={`${item?.id}-${index}`}>
                {
                  //@ts-ignore
                  <img className="img-fluid" src={item.urls?.regular} />
                }
              </div>
            ))}
          </div>
          {!loading && page < totalPages && (
            <div className="center load-more-wrapper">
              <button data-testid="load-more" onClick={onPaginationClick}>
                Load more
              </button>
            </div>
          )}
        </React.Fragment>
      )}
      {loading && <p className="center">Loading...</p>}
    </div>
  </React.Fragment>
);

export default Search;
