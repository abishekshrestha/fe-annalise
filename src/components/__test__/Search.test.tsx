import { fireEvent, render, screen } from "@testing-library/react";
import { data } from "../mocks/searchResult";
import Search from "../SecondPage/Search";

test("snapshot", () => {
  const renderedComponent = render(
    <Search
      name=""
      search=""
      imageList={[]}
      loading={false}
      isValid={false}
      totalPages={0}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  expect(renderedComponent).toMatchSnapshot();
});

test("show input error when isValid is false", () => {
  render(
    <Search
      name=""
      search=""
      imageList={[]}
      loading={false}
      isValid={false}
      totalPages={0}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const inputError = screen.getByText(/Please enter valid keyword/i);
  expect(inputError).toBeInTheDocument();
});

test("should not show input error when isValid is true", () => {
  render(
    <Search
      name=""
      search=""
      imageList={[]}
      loading={false}
      isValid={true}
      totalPages={0}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const inputError = screen.queryByText(/Please enter valid keyword/i);
  expect(inputError).toBeNull();
});

test("should show loading when loading is true", () => {
  render(
    <Search
      name=""
      search=""
      imageList={[]}
      loading={true}
      isValid={true}
      totalPages={0}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const loading = screen.queryByText("Loading...");
  expect(loading).toBeInTheDocument();
});

test("should show load more button when loading is false and totalPages is greater than page number", () => {
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={false}
      isValid={true}
      totalPages={5}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const loadMore = screen.getByRole("button", {
    name: /Load more/i,
  });
  expect(loadMore).toBeInTheDocument();
});

test("should not show load more button when loading is false and totalPages is equal to page number", () => {
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={false}
      isValid={true}
      totalPages={1}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const loadMore = screen.queryByRole("button", {
    name: /Load more/i,
  });
  expect(loadMore).toBeNull();
});

test("should not show load more button when loading is false and totalPages is less than page number", () => {
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={false}
      isValid={true}
      totalPages={2}
      page={3}
      onInputChange={jest.fn()}
    />
  );

  const loadMore = screen.queryByRole("button", {
    name: /Load more/i,
  });
  expect(loadMore).toBeNull();
});

test("should not show load more button when loading is true", () => {
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={true}
      isValid={true}
      totalPages={3}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const loadMore = screen.queryByRole("button", {
    name: /Load more/i,
  });
  expect(loadMore).toBeNull();
});

test("should have list of images from imageList prop", () => {
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={false}
      isValid={true}
      totalPages={0}
      page={1}
      onInputChange={jest.fn()}
    />
  );

  const imageList = screen.getByTestId("image-list");
  expect(imageList).toBeInTheDocument();
});

test("should submit search form", () => {
  const mockCallBack = jest.fn();
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={false}
      isValid={true}
      totalPages={0}
      page={1}
      onInputChange={jest.fn()}
      handleSubmit={mockCallBack}
    />
  );

  const formElement = screen.getByTestId("search-form");
  fireEvent.submit(formElement);

  expect(mockCallBack).toHaveBeenCalled();
});

test("should call onPaginationClick function on load more button click", () => {
  const mockCallBack = jest.fn();
  render(
    <Search
      name=""
      search=""
      imageList={data.results}
      loading={false}
      isValid={true}
      totalPages={5}
      page={1}
      onInputChange={jest.fn()}
      onPaginationClick={mockCallBack}
    />
  );

  const loadMoreButton = screen.getByTestId("load-more");
  fireEvent.click(loadMoreButton);

  expect(mockCallBack).toHaveBeenCalled();
});
