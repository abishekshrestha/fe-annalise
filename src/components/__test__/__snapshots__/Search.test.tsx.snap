// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header"
      >
        <div
          class="container"
        >
          <div
            class="row"
          >
            <h1
              class="col-12"
              id="name"
            >
              Hello, 
            </h1>
          </div>
        </div>
      </header>
      <div
        class="container image-list"
      >
        <div
          class="row"
        >
          <div
            class="col-12"
          >
            <form
              class="search-group"
              data-testid="search-form"
            >
              <label
                for="search"
                hidden=""
              >
                Search
              </label>
              <input
                data-testid="inputbox"
                id="search"
                name="search"
                placeholder="Search"
                type="text"
                value=""
              />
              <button
                type="submit"
              >
                Search
              </button>
            </form>
            <div
              class="error"
            >
              Please enter valid keyword
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <header
      class="header"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <h1
            class="col-12"
            id="name"
          >
            Hello, 
          </h1>
        </div>
      </div>
    </header>
    <div
      class="container image-list"
    >
      <div
        class="row"
      >
        <div
          class="col-12"
        >
          <form
            class="search-group"
            data-testid="search-form"
          >
            <label
              for="search"
              hidden=""
            >
              Search
            </label>
            <input
              data-testid="inputbox"
              id="search"
              name="search"
              placeholder="Search"
              type="text"
              value=""
            />
            <button
              type="submit"
            >
              Search
            </button>
          </form>
          <div
            class="error"
          >
            Please enter valid keyword
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
